extensions/common/android/xwalk_extension_android.cc

```diff
@@ -162,21 +178,33 @@ XWalkExtensionAndroidInstance::~XWalkExtensionAndroidInstance() {
 void XWalkExtensionAndroidInstance::HandleMessage(
     scoped_ptr<base::Value> msg) {
   std::string value;
-
-  if (!msg->GetAsString(&value)) {
-    return;
-  }
+  const base::BinaryValue* binary_value = nullptr;
 
   JNIEnv* env = base::android::AttachCurrentThread();
-  ScopedJavaLocalRef<jstring> buffer(env, env->NewStringUTF(value.c_str()));
   ScopedJavaLocalRef<jobject> obj = java_ref_.get(env);
   if (obj.is_null()) {
     LOG(ERROR) << "No valid Java object is referenced for message routing";
     return;
   }
 
-  Java_XWalkExtensionAndroid_onMessage(
-      env, obj.obj(), getID(), buffer.obj());
+  if (msg->GetAsString(&value)) {
+    ScopedJavaLocalRef<jstring> buffer(env, env->NewStringUTF(value.c_str()));
+    Java_XWalkExtensionAndroid_onMessage(
+        env, obj.obj(), getID(), buffer.obj());
+    return;
```

Comment:
All these new `return` statements aren't needed. You could rewrite this block like this:

```c++
if (GetAsString(...)) {
} else if (GetAsBinary(...) {
} else {
  NOTREACHED() << "Failed to decode message as either string or binary blob";
}
```

You could even replace the `NOTREACHED` with something more fatal if necessary; for example, is it possible that even decoding a binary blob fails?